README
=======================================================================
Release notes for E-Cell Simulation Environment Version 3 (E-Cell SE 3)
=======================================================================

Koichi Takahashi <shafi@e-cell.org>

E-Cell dev mailing list <dev@e-cell.org>

|build-status|

What is E-Cell System?
======================
E-Cell is a generic software package for cellular modeling and simulation.  

Files and directories
=====================
In the package, files are organized as follows::

  -+- README:    This file
   |
   +- INSTALL:   Installation instructions
   |
   +- ChangeLog: Automatically generated cvs log 
   |
   +- NEWS:      Project log
   |
   +- COPYING:   License
   |
   +- AUTHORS:   List of authors
   |
   +- ecell/ --> Core portion of E-Cell SE 3, or Hekkoro
   |  |
   |  +- libecs/       --> Koyurugi: Class library for cell modeling and
   |  |                    simulation (E-Cell SE 3's innermost core)
   |  +- libemc/       --> E-Cell Micro Core (EMC): C++ Facade API of Koyurugi
   |  |
   |  +- pyecs/        --> Python binding of EMC and Koyurugi
   |  |
   |  +- pyecell/      --> Python E-Cell library
   |  |
   |  +- session-monitor/  --> GUI session monitor (gecell)
   |  |
   |  +- model-editor/ --> GUI model editor (ecell3-modeleditor)
   |  |
   |  +- toollauncher/ --> GUI tool launcher 
   |
   +- dmtool/          --> C++ class library for dynamic loading of object
   |
   +- doc/             --> Documentations 
   |  |
   |  +- users-manual/ --> Users manual
   |  |
   |  +- samples/      --> Sample models
   |
   +- libltdl/         --> Libtool's dynamic module loader library

And installed in the following structure::

 PREFIX -+- bin/       --> E-Cell system toolsuite
         |
         +- include/   --> Header files (for developers)
         |  |
         |  +- dmtool/
         |  |
         |  +- ecell-3.1/
         |     |
         |     +- libecs/
         |     |
         |     +- libemc/
         |
         +- lib/       --> Shared libraries and python modules
         |  |
         |  +- python?.?/  --> Python modules
         |  |
         |  +- ecell-3.1/
         |     |
         |     +- dms      --> Standard dynamic modules
         |     |
         |     +- model-editor     --> Scripts and data for Model Editor
         |     |
         |     +- session-monitor  --> Scripts and data for Session Monitor
         |     |
         |     +- toollauncher     --> Scripts and data for Tool Launcher
         |
         +- share/
            |
            +- ecell-3.1/
            |  |
            |  +- dms/     --> Sources for the standard dynamic modules.
            |
            +- doc/
               |
               +- ecell/   --> E-Cell Documents
                  |
                  +- samples/      --> Sample models
                  |
                  +- users-manual* --> Users manual
                  |
                  +- api           --> API reference
                  |
                  +- model-editor  --> Model Editor documentation


About the code names
--------------------
Hekkoro
    Code name for E-Cell SE Version 3 (libecs, libemc and pyecs)
Koyurugi
    Code name for the libecs class library
Osogo
    Code name for the GUI Session monitor (gecell)

Documentation
=============
E-Cell 3 users manual is hosted on Read the Docs.
http://ecell3.readthedocs.org/en/latest/

Basic Usage
===========

1. Using E-Cell from python interpreter
---------------------------------------
By default a python binding (pyecs) is built.

To invoke a E-Cell Session Script interpreter in interactive mode, use ``ecell3-sesion`` script. Note that the script expects {PREFIX}/bin is included in PATH environment, where {PREFIX} is the directory path that was specified to configure. 
(If you have not specified it, it falls back to ``/usr/local``.)

A typical command-line session is as follows::

   $ ecell3-session
   ecell3-session [ E-Cell SE Version 3.1.106, on Python Version 2.3.4 ]
   Copyright (C) 1996-2020 Keio University
   Copyright (C) 2008-2020 RIKEN
   Copyright (C) 2005-2009 The Molecular Sciences Institute
   Send feedback to Koichi Takahashi <shafi@e-cell.org>
   ecell3-session>>>

In the above output, ``ecell3-session>>>`` is the prompt of ecell3-session. That accepts almost everything of interactive Python.

The below example first loads a model that is stored in simple.eml, and then run it for 10 seconds in simulation time::

   ecell3-session>>> loadModel( 'simple.eml' )
   <simple.eml, t=0>>> run( 10 )
   <simple.eml, t=10.001>>>

2. Using E-Cell from Session Monitor GUI
----------------------------------------
If you have python, and pygtk2 installed, a GUI session monitor, is built.

To use Session Monitor, type::

   $ ecell3-session-monitor

Sample Models
=============

Sample models are installed under {PREFIX}/share/doc/samples.  Each model is
stored as a EM script, while ecell toolsuite only accept XML-based EML files.

EM scripts can be converted to EML files through the program ``ecell3-em2eml``::

   $ ecell3-em2eml Drosophila.em

This will create Drosophila.eml in the current working directory.

Some samples require that the accompanying dynamic modules (DM's) be
built before running simulation.  These modules are provided as C++ sources
which can be compiled using ``ecell3-dmc``::

   $ ecell3-dmc SomeModule.cpp

For more information about the dynamic module, refer to the users manual.


Copyright Notice
================
E-Cell System is distributed under the GNU General Public License
version 2.

  Copyright (C) 1996-2020 Keio University
  Copyright (C) 2008-2020 RIKEN
  Copyright (C) 2005-2009 The Molecular Sciences Institute

  E-Cell System is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public
  License as published by the Free Software Foundation; either
  version 2 of the License, or (at your option) any later version.
  
  E-Cell System is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  See the GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public
  License along with E-Cell System -- see the file COPYING.
  If not, write to the Free Software Foundation, Inc.,
  59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 
It should be mentioned this package contains the following softwares
for convenience.  Please note that these are not covered by the license
under which a copy of E-Cell System is licensed to you, while neither
composition nor distribution of any derivative work of E-Cell System with
these software violates the terms of each license, provided that it meets
every condition of the respective licenses.

* PLY

  PLY is written by David M. Beazley and distributed under the GNU
  Lesser General Public License.  See COPYING.LGPLv2 or COPYING.LGPLv3 for the
  full license terms of the LGPL(s).

    Author: David M. Beazley (beazley@cs.uchicago.edu)
            Department of Computer Science
            University of Chicago
            Chicago, IL  60637
    
    Copyright (C) 2001, David M. Beazley
    
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.
    
    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.
    
    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

  The website for this software is located at:

    http://www.dabeaz.com/ply/

* EmPy 

  EmPy is written by Erik Max Francis and distributed under the GNU Lesser
  General Public License.  See COPYING.LGPLv2 or COPYING.LGPLv3 for the full
  license terms of the LGPL(s).

    Copyright (C) 2002-2003 Erik Max Francis <max@alcyone.com>

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.
    
    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.
    
    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

  The website for this software is located at:

    http://www.alcyone.com/software/empy/

* libltdl

  libltdl is part of libtool, whose copyright is held by Free Software
  Foundation.  See COPYING.LGPLv2 or COPYING.LGPLv3 for the full license terms
  of the LGPL(s).

    Copyright (C) 1998, 1999, 2000 Free Software Foundation, Inc.
    Originally by Thomas Tanner <tanner@ffii.org>

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.
     
    As a special exception to the GNU Lesser General Public License,
    if you distribute this file as part of a program or library that
    is built using GNU libtool, you may include it under the same
    distribution terms that you use for the rest of that program.
       
    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

  The website for this software is located at:

    http://www.gnu.org/software/libtool/

* The Loki Library

  Libecs contains part of The Loki Library, which is writen by Andrei 
  Alexandrescu and the contributors.

    Copyright (c) 2001 by Andrei Alexandrescu

    This code accompanies the book:
    Alexandrescu, Andrei. "Modern C++ Design: Generic Programming and Design 
    Patterns Applied". Copyright (c) 2001. Addison-Wesley.

    Permission to use, copy, modify, distribute and sell this software for any 
    purpose is hereby granted without fee, provided that the above copyright 
    notice appear in all copies and that both that copyright notice and this 
    permission notice appear in supporting documentation.

    The author or Addison-Wesley Longman make no representations about the 
    suitability of this software for any purpose. It is provided "as is" 
    without express or implied warranty.

  The website for this software is located at:

    http://loki-lib.sourceforge.net/

* TableIO

   Table IO is a module for reading ASCII tables from files to Python lists
   written by Michael A. Miller and distributed under the GNU General Public
   License.  See COPYING.GPLv2 or COPYING.GPLv3 for the full license terms of
   the GPL(s).

     Copyright (C) 2000 Michael A. Miller <mmiller@debian.org>

     This program is free software; you can redistribute it and/or
     modify it under the terms of the GNU General Public License as
     published by the Free Software Foundation; either version 2 of the
     License, or (at your option) any later version.
     
     This program is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.
     
     You should have received a copy of the GNU General Public License
     along with this program; if not, write to the Free Software
     Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
     USA. 

  The website for this software is located at:

     http://php.iupui.edu/~mmiller3/python/

.. Build status badge
.. |build-status|
   image:: https://secure.travis-ci.org/ecell/ecell3.png
   :target: http://travis-ci.org/ecell/ecell3
   :alt: Build Status

$Id$



NEWS
ecell-3.2.8: 26-Oct-2017
 * Remove PythonProcesses from make targets.
 * Add ecell3-jupyter that is idential to ecell3-ipython.
 * Improve setModel() (now second arg is omittable).
 * Fix a bug in the Python module ecell.FullID.

ecell-3.2.7: 03-Nov-2017
 * Fixed a bug in the Python module ecell.FullID.

ecell-3.2.6: 25-Oct-2017
 * ecell3-model-editor is removed from build targets.
 * Fixed a bug.

ecell-3.2.5: 18-Oct-2017
 * ecell3-session: word completion function is implemented.
 * Add new samples
 * Fixed bugs

ecell-3.2.4d: 05-Jun-2017
 * Fix an ecell3-session bug that destroy the readline function after load/setModel().
 * Add three new samples: ga-deap, SRES, steady_state
 * Add an alias to Python API: getClassName() -> getClasname()

ecell-3.2.4c: 12-May-2017

 * Improve setModel() (add EML support)

 * Update samples/simple

ecell-3.2.4b: 11-May-2017

 * Add setModel()


ecell-3.2.4: 4-May-2017

 * Fixed a long standing bug in ODE23Stepper and ODE45Stepper.

 * Add Python System.configureSize() method.

 * Add Model::Model::createSystemEntity()

 * Support Torque scheduler.


ecell-3.2.3pre2: 21-Jan-2012

 * Fixed a bug that EntityStub.setProperty() doesn't work since 3.2.3pre1.

 * Added workaround for Apple's GCC extensions.

 * Made ga.py work with the new SessionManager.


ecell-3.2.3pre1: 28-Oct-2011

 * Made the simulator object available to the expressions in PythonProcess.

 * Made session_monitor reloadable.

 * Added missing accessors in Python binding.

 * Supported Boost >= 1.43.0

 * Made compatible Better compatibility with newer


ecell-3.2.2: 5-Jan-2011

 * Fixed the bug that some expressions given to ExpressionFluxProcess or
   ExpressionAssignmentProcess would not be evaluated correctly.

 * Fixed segmentation fault bug in ExpressionAssignmentProcess.

 * Overhauled libecs::Polymorph.

 * Fixed a bug in ecell.session_manager that SGE backend doesn't work
   unless setOwner() is called prior to job dispatch.

 * Fixed compilation error in ecell/libecs/MethodProxy.hpp.

 * Fixed ecell._ecs module so that iter(process.variableReferences) works
   correctly.

 * Fixed the bug that logger data could not be saved through LoggerWindow.


ecell-3.2.1: 23-Sep-2010

 * Fixed segmentation fault in PythonProcess and PythonFluxProcess.

 * Removed DifferentialStepper::initializeStepInterval() in favor of
   virtualized Stepper::setStepInterval().

 * Better compatibility with Fujitsu C++ compiler.

 * Fixed possible segmentation fault that occurs during loading a DM.

 * Removed funky macros derived from UtilLinux.

 * Optimized code in VariableReference.


ecell-3.2.0p2: 2-Apr-2010

 * Fixed DataPointVector's python wrapper to have "shape" property that returns
   the shape of the array.  That is required by TableIO module.
   (notified by Erin Rachael Shellman. Thanks.)


ecell-3.2.0p1: 9-Feb-2010

 * Fixed the build system so that resource files for the GUI tools are
   correctly installed under the package data directory.

 * Fixed FileSelection combatibility wrapper so that they can present
   the text box for the filename entry.  Because of this bug, models could not
   be saved in session-monitor or model editor.

 * Made ecell3-dmc and dmcompile available on Windows.

 * Fixed a bug in System::detach() that causes random segfault on cleanup.

 * Fixed infinite recursion in ecell._ecs.Simulator.getDMSearchPath().


ecell-3.2.0: 5-Feb-2010

 * See the entries for 3.1.108.


ecell-3.2.0RC4: 3-Feb-2010

 * Fixed the release dates in NEWS file.

 * Removed System::dispose() that used to do the same clean-up process as
   System::detach() that is eventually called via EcsObject::dispose().
   This caused long-standing problem on Windows with MSVC.

 * Fixed typo in the MSVC build file for pyecell.


ecell-3.2.0RC3: 27-Jan-2010

 * Modified libecs::Stepper::log() so that it can be compiled with older
   version of Boost.

 * Added our version of std::bind2nd() to Utils.hpp for better compatibility.

 * For the rest of the changes, see the entries for 3.1.108RC3.


ecell-3.2.0RC2: 16-Jan-2010

 * See the entries for 3.1.108RC2


ecell-3.2.0RC1: 13-Jan-2009

 * A few cosmetic changes

   - Fixed indent / coding style.

   - Fixed doxygen comments.

   - Removed $...$ placeholders as they may contain localized characters
     which are troublesome in VC++.

 * dmtool changes

   - ModuleMaker class is removed and StaticModuleMaker<> template class is
     renamed to ModuleMaker<>.

   - Added SharedModuleMakerBase class that is the base class of
     SharedModuleMaker<> template classes.  Initialization and finalization
     static methods are moved from ModuleMaker to it.

   - SharedModuleMakerBase provides a couple of instance methods,
     setSearchPath() and getSearchPath() which sets or gets the path used when.
     looking up for a shared module.

   - Purged SharedDynamicModule and SharedModuleMaker into
     SharedDynamicModule.hpp and SharedModuleMaker.hpp respectively, in order
     to remove unwanted dependencies on libltdl.

 * libecs changes

   - Thoroughly overhauled DM/PropertiedClass ABI. Backwards binary
     compatibility was lost.

     - Added DynamicModuleDescriptor class that is defined per one dynamic
       module, and retains the information and a few set of fucntion pointers
       that are required to instantiate the module.

     - Added DynamicModuleInfo interface by which the meta-information
       regarding the dynamic module can be retrieved.

     - <AnyDMObject>::getClassInfo() now returns the PropertyInterface instead
       of synthesized Polymorph map.

     - PropertyInterface class is now a subclass of DynamicModuleInfo.

     - PropertyInterface objects are now initialized from within the associated
       ModuleMaker's rather than by the static initializers.

     - Removed PropertyInterface::getPropertyList().

     - Added PropertyInterface::getPropertyAttributes() that can be used in

ecell/libecs/osif.h
//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
//
//       This file is part of the E-Cell System
//
//       Copyright (C) 1996-2019 Keio University
//       Copyright (C) 2008-2019 RIKEN
//       Copyright (C) 2005-2009 The Molecular Sciences Institute
//
//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
//
//
// E-Cell System is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public
// License as published by the Free Software Foundation; either
// version 2 of the License, or (at your option) any later version.
// 
// E-Cell System is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public
// License along with E-Cell System -- see the file COPYING.
// If not, write to the Free Software Foundation, Inc.,
// 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
// 
//END_HEADER
//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
//	This file is a part of E-CELL2.
//	Original codes of E-CELL1 core were written by Koichi TAKAHASHI
//	<shafi@e-cell.org>.
//	Some codes of E-CELL2 core are minor changed from E-CELL1
//	by Naota ISHIKAWA <naota@mag.keio.ac.jp>.
//	Other codes of E-CELL2 core and all of E-CELL2 UIMAN are newly
//	written by Naota ISHIKAWA.
//	All codes of E-CELL2 GUI are written by
//	Mitsui Knowledge Industry Co., Ltd. <http://bio.mki.co.jp/>
//
//	Latest version is availabe on <http://bioinformatics.org/>
//	and/or <http://www.e-cell.org/>.
//END_V2_HEADER
//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
/*
 *::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
 *	$Id$
 :	$Log$
 :	Revision 1.2  2005/11/19 09:23:59  shafi
 :	Kouichi -> Koichi
 :
 :	Revision 1.1  2002/04/30 11:21:53  shafi
 :	gabor's vvector logger patch + modifications by shafi
 :	
 :	Revision 1.5  2002/01/15 14:53:43  ishikawa
 :	osif_add_path()
 :	
 :	Revision 1.4  2001/10/21 15:27:12  ishikawa
 :	osif_is_dir()
 :	
 :	Revision 1.3  2001/03/23 18:51:17  naota
 :	comment for credit
 :	
 :	Revision 1.2  2001/01/13 01:31:47  naota
 :	Can be compiled by VC, but does not run.
 :
 :	Revision 1.1  2000/12/30 15:09:46  naota
 :	Initial revision
 :
//END_RCS_HEADER
 *::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
 */
/*
 * OS dependent module
 * Written by ISHIKAWA Naota.  (C) 2000 Keio university
 */
#ifdef __cplusplus
extern "C" {
#endif
#ifndef __OSIF_H__
#define __OSIF_H__ 1

#ifdef WIN32
#include "unistd.h"
#else
#include <sys/types.h>
#endif


long osif_get_pid();
off_t osif_disk_free(const char *__path);	/* by K Bytes */
int osif_mkdir(const char *__name);
int osif_is_dir(const char *__name);
int osif_load_dll(const char *__name);
int osif_add_path(const char *__path, int to_first);

#ifdef __cplusplus
} /* end of extern "C" */
#endif /* __cplusplus */
#endif /* __OSIF_H__ */
